torchtune/datasets/multimodal/_the_cauldron.py

```diff
@@ -0,0 +1,165 @@
+                    content=[
+                        {"type": "image"},
+                        {"type": "text", "content": message["user"]},
+                    ],
+                    masked=not self.train_on_input,
+                )
+            )
+            messages.append(
+                Message(
+                    role="assistant",
+                    content=[{"type": "text", "content": message["assistant"]}],
+                )
+            )
+
+        if self.new_system_prompt is not None:
+            messages = [
+                Message(
+                    role="system", content=self.new_system_prompt, masked=True, eot=True
+                )
+            ] + messages
+
+        return {"messages": messages, "images": sample[self._column_map["images"]]}
+
+
+def the_cauldron_dataset(
+    model_transform: Transform,
+    *,
+    subset: str,
+    source: str = "HuggingFaceM4/the_cauldron",
+    column_map: Optional[Dict[str, str]] = None,
+    new_system_prompt: Optional[str] = None,
+    train_on_input: bool = True,
+    packed: bool = False,
+    split: str = "train",
+    **load_dataset_kwargs: Dict[str, Any],
+) -> Union[SFTDataset, PackedDataset]:
+    """
+    Support for family of image + text datasets similar to
+    `The Cauldron<https://huggingface.co/datasets/HuggingFaceM4/the_cauldron>_`
+    from Hugging Face Datasets.
+
+    Args:
+        model_transform (Transform): model-specific transform that takes in a sample dict and applies custom
+            transforms on the keys. The tokenizer used by the model should be encapsulated in the model transform
+            and should operate on the "messages" field. Any model-specific image transforms should operate on
+            the "images" field. The keys returned by the model should be aligned with the
+            expected inputs into the model.
+        subset (str): name of the subset of the dataset to load. See the `dataset card
+            <https://huggingface.co/datasets/HuggingFaceM4/the_cauldron>`_ for options.
+        source (str): path to dataset repository on Hugging Face. For local datasets,
+            define source as the data file type (e.g. "json", "csv", "text") and pass
+            in the filepath in ``data_files``. See `Hugging Face's
+            <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path>`_
+            ``load_dataset`` for more details. Default is ``HuggingFaceM4/the_cauldron``.
+        column_map (Optional[Dict[str, str]]): a mapping to change the expected "images"
+            and "texts" column names to the actual column names in the dataset. Default is None,
+            keeping the default column names.
+        new_system_prompt (Optional[str]): if specified, prepend a system message. This can
+            serve as instructions to guide the model response. Setting this will OVERRIDE any system
+            messages already present in the dataset. Default is None.
```

Comment:
I think I missed the original PR here, but this naming feels a bit unclear to me (though I guess the purpose of "new" is to emphasize that we're overriding any other system prompts). Also again would give an example of this

Reply:
Yep that's exactly it